### PR TITLE
Warn if layout doesn't stabilize

### PR DESCRIPTION
**Pull request by bluebear94**
This adds a warning if layout does not converge in the allotted number of iterations.

An example of a warning raised with this flag:

```
warning: layout did not converge in 5 attempts
  ┌─ /loop.typ:1:0
  │  
1 │ ╭ #let s = state("x", 1)
2 │ │ #locate(loc => {
3 │ │   s.update(s.final(loc) + 1)
4 │ │ })
5 │ │ #s.display()
6 │ │ 
  │ ╰^
  │  
  = hint: check if any states or queries are updating themselves
```

I also considered adding an option to control the maximum number of iterations itself, but @laurmaedje was against it.

### CODE REVIEW

**Comment by laurmaedje** [comment]
Maybe I wasn't clear enough, but I also do not like this flag. But I'm fine with throwing this error unconditionally.

**Comment by bluebear94** [comment]
Ideally, I’d like to make this situation warn by default, but it seems that it’s currently not possible to issue warnings.

**Comment by PgBiel** [comment]
> But I'm fine with throwing this error unconditionally.

I think throwing this error unconditionally is a bad idea; it might not be the user's fault that this is happening (e.g. could be some bad library or a problem in Typst itself), and the user would be frustrated to be unable to fix it and thus not be able to work on their document. Of course, the document's output wouldn't be ideal in such a case, but it's better than nothing.

I think having this be a (configurable) warning or something like that would be better in general.

**Comment by laurmaedje** [comment]
Then let's wait until we have warnings and then we'll make in an unconfigurable warning.

**Comment by laurmaedje** [comment]
Warnings are implemented now!

**Comment by laurmaedje** [comment]
The delayed list shouldn't contain warnings I think. It's only delayed to prevent it from being fatal -- which warnings aren't by definition.

**Comment by PgBiel** [comment]
Regarding the `--require-layout-converge` flag (which was, imo rightfully, removed), we should probably leave that for a later and more general feature of warning suppression.

File: crates/typst/src/model/mod.rs [hunk 1]
```diff
@@ -82,7 +82,20 @@ pub fn typeset(
...
+
+        if iter >= 5 {
+            tracer.warn(
+                SourceDiagnostic::warning(
+                    world.main().root().span(),
+                    "layout did not converge in 5 attempts",
+                )
+                .with_hints([
+                    "check if any states or queries are updating themselves".into(),
+                ]),
```
Comment (laurmaedje) [inline crates/typst/src/model/mod.rs:97]:
```suggestion
                warning!(
                    world.main().root().span(),
                    "layout did not converge within 5 attempts",
                )
                .with_hint(
                    "check if any states or queries are updating themselves".into(),
                ),
```

**Comment by laurmaedje** [comment]
I made one minor code style suggestion. Are we happy to merge this or do we want to wait until a warning suppression system is in place?

**Comment by bluebear94** [comment]
If we later add a warning suppression system, then we can always revisit this piece of code to add the necessary code to make it work with it.

**Comment by laurmaedje** [comment]
Sure, I more meant because @PgBiel express the concern of adding too many warnings before you can disable them. But this particular one is very reasonable to show, so it should be fine.

**Comment by laurmaedje** [comment]
Tests failed because of my slight rewording of the message. Sorry about that, didn't think of that or else I had made a suggestion on the test, too.

**Comment by laurmaedje** [comment]
Thanks!

**Pull request merged**